File: app/lib/features/product/models/review_response.dart
import 'package:json_annotation/json_annotation.dart';

part 'review_response.g.dart';

/// A single product review (dto-spec §16). The client shows an "edited"
/// indicator when [updatedAt] differs from [createdAt].
///
/// [productId] links the review to its product and [userId] is the author's
/// user id; together they let the client recognise the caller's own review
/// (`review.userId == currentUser.id`) to offer Edit / Delete. Authorization
/// stays server-enforced.
@JsonSerializable(createToJson: false)
class ReviewResponse {
  const ReviewResponse({
    required this.id,
    required this.productId,
    required this.userId,
    required this.userFullName,
    required this.rating,
    required this.createdAt,
    required this.updatedAt,
    this.userAvatarUrl,
    this.comment,
  });

  factory ReviewResponse.fromJson(Map<String, dynamic> json) =>
      _$ReviewResponseFromJson(json);

  final int id;

  /// The reviewed product's id.
  final int productId;

  /// The review author's user id.
  final int userId;
  final String userFullName;
  final String? userAvatarUrl;

  /// 1–5.
  final int rating;
  final String? comment;
  final DateTime createdAt;
  final DateTime updatedAt;
}

File: src/main/java/com/footverse/review/dto/ReviewResponse.java
package com.footverse.review.dto;

import java.time.LocalDateTime;

/**
 * A single product review returned to any reader (dto-spec §16). The author is exposed only through
 * display fields ({@code userFullName}, {@code userAvatarUrl}); the underlying {@code User} entity
 * is never serialized. The client shows an "edited" indicator when {@code updatedAt} differs from
 * {@code createdAt}.
 *
 * <p>{@code productId} links the review to its product and {@code userId} is the author's user id;
 * together they let the client recognise the caller's own review ({@code review.userId ==
 * currentUser.id}) and offer Edit / Delete. Both are plain identifiers carried straight from the
 * persisted review — authorization stays server-enforced (dto-spec §16).</p>
 *
 * @param id            the review id
 * @param productId     the reviewed product's id
 * @param userId        the review author's user id
 * @param userFullName  the author's full name
 * @param userAvatarUrl the author's avatar URL, if any
 * @param rating        the rating, 1–5
 * @param comment       the optional comment (max 500 chars), if any
 * @param createdAt     the creation timestamp
 * @param updatedAt     the last-update timestamp
 */
public record ReviewResponse(
        Long id,
        Long productId,
        Long userId,
        String userFullName,
        String userAvatarUrl,
        int rating,
        String comment,
        LocalDateTime createdAt,
        LocalDateTime updatedAt) {
}
